src/radical/pilot/agent/executing/popen.py

```diff
@@ -417,73 +437,71 @@ def _watch(self):
     # --------------------------------------------------------------------------
     # Iterate over all running tasks, check their status, and decide on the
     # next step.  Also check for a requested cancellation for the tasks.
-    def _check_running(self):
+    def _check_running(self, to_watch, to_cancel):
+
+        action = False
+        for task in to_watch:
```

Comment:
`for task in list(to_watch):`? (or some other iteration over `to_watch` list)

just found it as a long standing bug - some tasks could be skipped until the following round(s), due to other tasks are deleted during the iteration process - gives some delay for tasks processing

```
>>> tasks = [{1:1}, {2:2}, {3:3}]
>>> for task in tasks:
...     print('Current task: %s (from %s)' % (task, tasks))
...     if 1 in task or 2 in task:
...         tasks.remove(task)
... 
Current task: {1: 1} (from [{1: 1}, {2: 2}, {3: 3}])
Current task: {3: 3} (from [{2: 2}, {3: 3}])
```